browser-sign-in/__tests__/App-test.js

```diff
@@ -1,12 +1,205 @@
+    const getUserFromIdButton = wrapper.find('Button').get(1);
+    const getUserButton = wrapper.find('Button').get(2);
+    const getUserFromTokenButton = wrapper.find('Button').get(3);
+    expect.assertions(4);
+    expect(logoutButton.props.title).toBe('Logout');
+    expect(getUserFromIdButton.props.title).toBe('Get User From Id Token');
+    expect(getUserButton.props.title).toBe('Get User From Request');
+    expect(getUserFromTokenButton.props.title).toBe('Get User From Access Token');
+  });
+
+  it('should not render login button if authenticated', () => {
+    const wrapper = shallow(<App />);
+    wrapper.setState({authenticated: true});
+    const loginButton = wrapper.find('Button').get(0);
+    expect.assertions(1);
+    expect(loginButton.props.title).not.toBe('Login');
+  });
+
+  it('should not render logout and get user info buttons if not authenticated', () => {
+    const wrapper = shallow(<App />);
+    const logoutButton = wrapper.find('Button').get(0);
+    const getUserFromIdButton = wrapper.find('Button').get(1);
+    const getUserButton = wrapper.find('Button').get(2);
+    const getUserFromTokenButton = wrapper.find('Button').get(3);
+    expect.assertions(4);
+    expect(logoutButton.props.title).not.toBe('Logout');
+    expect(getUserFromIdButton).toBe(undefined);
+    expect(getUserButton).toBe(undefined);
+    expect(getUserFromTokenButton).toBe(undefined);
+  });
+});
+
+describe('when authenticated', () => {
+  it('should detect when the user has logged in', async () => {
+    const wrapper = shallow(<App />);
+    const loginButton = wrapper.find('Button').get(0);
+    expect.assertions(3);
+    await loginButton.props.onPress();
+    expect(loginButton.props.title).toBe('Login');
+    nativeEmitter.emit('signInSuccess');
+    expect(wrapper.state().authenticated).toBe(true);
+    expect(wrapper.state().context).toBe('Logged in!');
+  });
+
+  it('should detect when the user has signed out', async () => {
+    const wrapper = shallow(<App />);
+    wrapper.setState({authenticated: true});
+    const logoutButton = wrapper.find('Button').get(0);
+    expect.assertions(3);
+    await logoutButton.props.onPress();
+    expect(logoutButton.props.title).toBe('Logout');
+    nativeEmitter.emit('signOutSuccess');
+    expect(wrapper.state().authenticated).toBe(false);
+    expect(wrapper.state().context).toBe('Logged out!');
+  });
+
+  it('should return user profile information from id token' , async () => {
+    const mockGetIdToken = require('react-native').NativeModules.OktaSdkBridge.getIdToken;
+    mockGetIdToken.mockImplementationOnce(() => {
+      return {'id_token': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhIjoiYiJ9.jiMyrsmD8AoHWeQgmxZ5yq8z0lXS67_QGs52AzC8Ru8'};
+    });
+    const wrapper = shallow(<App />);
+    wrapper.setState({authenticated: true});
+    const profileButton = wrapper.find('Button').get(1);
+    await profileButton.props.onPress();
+    setTimeout( function() { 
```

Comment:
In react, setState() is async, and I believe wrapper.setState is the same here -

You can pass a function as the second param to setState, and it will be called when the setState is complete.  Given your collection of tasks here, that can get a little dicey (like, you might want to create a promise and pass the resolve as the callback to setState, then put the rest in the then() clause).

You can see more here: https://stackoverflow.com/questions/50667197/reactjest-testing-async-components-and-waiting-for-mount

Reply:
Removed setTimeout, and changes to using waitForState with enzyme async helper